napari/_app_model/context/_context.py

```diff
@@ -16,6 +17,32 @@
 __all__ = ['create_context', 'get_context', 'Context', 'SettingsAwareContext']
 
 
+class ContextMapping(collections.abc.Mapping):
+    def __init__(self, initial_values: collections.abc.Mapping):
```

Comment:
Need docstring, I would suggest something like:

```python
"""
This object wraps around app-model contexts and enables functional context-keys that are only evaluated
when they're first queried. `ContextMapping` objects are created from a context any time someone calls
`NapariApplication.get_context`. This usually happens just before a menu is about to be shown, when we
update the menu's actions' states based on the values of the context keys. The call to `get_context`
triggers the creation of the `ContextMapping` which stores (or, in the case of functional keys, evaluates
then stores) the value of each context key. Once keys are evaluated, they are cached within the object for
future accessing of the same keys. However, any new `get_context` calls will create a brand new
`ContextMapping` object. 
"""
```

Reply:
This will probs require docs updates as well, @lucyleeow do you know specific spots where this will need changing in the docs? Also what do you think of this docstring @jni, @Czaki, @lucyleeow

Reply:
Hmm I still don't love that we create is new context every time we `get_context` Is there a way for it to persist and allow us to invalidate stuff (as described in: https://github.com/napari/napari/pull/6965#pullrequestreview-2111877494)

Also I prefer the docstring convention of having a one liner followed by longer paragraph(s).

Our context doc is here: https://napari.org/dev/guides/contexts_expressions.html# BUT it is out of date since this moved to app-model - app-model have their own context docs, which I think we should refer to instead in some cases (see https://github.com/napari/docs/issues/369 and https://github.com/napari/docs/issues/146). But this doc update is probably not trivial.

Reply:
> Hmm I still don't love that we create is new context every time we `get_context` Is there a way for it to persist and allow us to invalidate stuff (as described in: [#6965 (review)](https://github.com/napari/napari/pull/6965#pullrequestreview-2111877494))

Every option that I could design currently is more expensive in cost of calculation than recreate as it will require to design complex querying "when to invalidate"

Reply:
I am not sure what you mean by 'complex querying'. I thought the intention was to have these context functions have a lru cache and to connect 'invalidating' events to a clear cache function? I can see that this would be more complex and haven't thought about implementation details, so not clear it would work well.

If we don't want to bother with something more complex (which would be understandable) I would not emphasis (or even do away with all together): the once only evaluation and they value caching. If we create a new context instance on `get_context`, which gets called before we update the context, I don't see a way where the above functionalities would actually ever be used?

If we find that we want to use this functionality later we can always add it?

Reply:
Why don't we just make a simple class that evaluates functions, because thats all we are using anyway?

Reply:
Ignore that last part! I've realised that we may query more than once in the update. 🤦

Reply:
**Functions / cache**

> Functions are for things where is no event.

So in the two cases where we currently have a function context key in napari, I would argue that there is/could be an event, BUT I agree that it may not be worth the extra complexity/impact performance to use it. E.g.,

* 'valid_spatial_json_clipboard' - invalidate cache could be connected when the clipboard changes
* 'is_set_trace_active' - invalidate cache could be connected to when 'trace_file' changes on the timer (we'd have to add such an event)

> If we create a new context instance on get_context, which gets called before we update the context, I don't see a way where the above functionalities would actually ever be used?

Ignore this comment, I got confused and forgot that when we `update_from_context` we will query the context map.

Reply:
There is a third function `_empty_shapes_layer_selected` that is resolving #6643, that points to a class of problems of updating inner layer state and complex connection/disconnection from signals. 

Also, `ContextMapping` allows delaying calculation of some more problematic context keys until they are needed.

Reply:
Ah I see, that is clever. I think there are some problems with this implementation but I cannot think of anything better so I'm happy to go with this and change in future if need be.

I think we do need to update the documentation (it is at least important for devs to know this), and make it clear that:

* context keys connected to events are updated immediately, and will always have the 'correct' value
* function context keys only get evaluated when we `get_context`, this is why we create a new context instance and update from context on `aboutToShow`.

(this also raises the question of whether it is a good idea to have context keys behave these two different ways. Obviously, depending on event vs menu click frequency, performance will be better one way or the other - BUT we are not splitting the behaviour of context keys depending on what will give us best performance, we're just doing what's easiest at the time implementation wise :sweat_smile: ) 

Also I kept #6643 open because we needed to create a general issue, as there are a fair number of context keys that change with data changes, see: https://github.com/napari/napari/issues/6643#issuecomment-1943212664. I will create a new issue - and we would need to amend all data related context keys to be functions to fix this issue.

Reply:
Ill make an issue over in docs detailing this to remind us to document it